### PR TITLE
[Core] Fix segfault when cancel and generator is used with a high loa…

### PR DESCRIPTION
**Pull request by rkooo567**
…d #40083 (#40122)

## Checks

- [ ] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
    - [ ] I've added any new APIs to the API Reference. For example, if I added a 
           method in Tune, I've added it in `doc/source/tune/api/` under the 
           corresponding `.rst` file.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

**Review by zhe-thoughts — APPROVED** [review]

File: python/ray/tests/test_streaming_generator_3.py [hunk 1]
```diff
@@ -231,6 +238,93 @@ def g(sleep_time):
     assert result[10] == 4
 
 
+def test_streaming_generator_load(shutdown_only):
```
Comment (sihanwang41) [inline python/ray/tests/test_streaming_generator_3.py:241]: is it deterministic to fail without this change? I was a little concern we have to send huge mount of traffic to trigger this issue.

Comment (edoakes) [reply python/ray/tests/test_streaming_generator_3.py:241]: Every time I ran this test locally w/o the change it segfaulted. W/ the change it never did.

**Review by edoakes — APPROVED** [review]

File: python/ray/tests/test_streaming_generator_3.py (hunk 1 above)
Comment (sihanwang41) [reply python/ray/tests/test_streaming_generator_3.py:241]: what do you think we move this test to serve unit test? @edoakes @rkooo567

Comment (sihanwang41) [reply python/ray/tests/test_streaming_generator_3.py:241]: if there is no pure core repro.

**Pull request merged**